File: Cpp/257.cpp
#include<iostream>
#include<vector>

#include "Tree.h"

using namespace std;

//recursive
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        string path;
        preorder(root, path, res);
        return res;
    }
    
    void preorder(TreeNode* root, string path, vector<string>& res){
        if(root){
            if(!root->left && !root->right){
                res.push_back(path+to_string(root->val));
                return;
            }
            
            preorder(root->left, path+to_string(root->val)+"->", res);
            preorder(root->right, path+to_string(root->val)+"->", res);
        }
    }
};

//iterative
class Solution2 {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        queue<string> paths;
        queue<TreeNode*> nodes;
        vector<string> res;
        if(root){
            nodes.push(root);
            paths.push(to_string(root->val));
        }
        
        while(!nodes.empty()){
            int size = nodes.size();
            while(size--){
                TreeNode* cur_node = nodes.front();
                nodes.pop();
                string cur_path = paths.front();
                paths.pop();
                
                if(!cur_node->left && !cur_node->right){
                    res.push_back(cur_path);
                    continue;
                }
                
                if(cur_node->left){
                    nodes.push(cur_node->left);
                    paths.push(cur_path+"->"+to_string(cur_node->left->val));
                }
                if(cur_node->right){
                    nodes.push(cur_node->right);
                    paths.push(cur_path+"->"+to_string(cur_node->right->val));
                }
            }
        }
        
        return res;
    }
};

int main(){

}

File: Python/257.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        res = []
        self.dfs(root, res, "")
        return res
    
    def dfs(self, root, res, path):
        if not root:
            return
        if not root.left and not root.right:
            path += str(root.val)
            res.append(path)
            return
        self.dfs(root.left, res, path + str(root.val) + "->")
        self.dfs(root.right, res, path + str(root.val) + "->")
